### PR TITLE
Further reduce usages of `StringUtils`

### PR DESCRIPTION
**Pull request by BobDu**
like #6270 

### Proposed changelog entries

- JENKINS-XXXXX, human-readable text

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: cli/src/main/java/hudson/cli/CLI.java [hunk 1]
```diff
@@ -246,7 +245,7 @@ public static int _main(String[] _args) throws Exception {
         if (auth == null && bearer == null) {
             // -auth option not set
             if ((userIdEnv != null && !userIdEnv.isBlank()) && (tokenEnv != null && !tokenEnv.isBlank())) {
-                auth = StringUtils.defaultString(userIdEnv).concat(":").concat(StringUtils.defaultString(tokenEnv));
+                auth = userIdEnv.concat(":").concat(tokenEnv);
```
Comment (StefanSpieker) [inline cli/src/main/java/hudson/cli/CLI.java:248]: I think you should use `Objects.toString(s, "");` for `StringUtils.defaultString(s);` as stated [here](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#defaultString-java.lang.String-java.lang.String-) which is also deprecated in commons lang 3

```suggestion
                auth = Objects.toString(userIdEnv, "").concat(":").concat(Objects.toString(tokenEnv, ""));
```

File: core/src/main/java/hudson/model/Slave.java [hunk 2]
```diff
@@ -393,7 +392,7 @@ public FilePath getRootPath() {
             // if computer is null then channel is null and thus we were going to return null anyway
             return null;
         } else {
-            return createPath(StringUtils.defaultString(computer.getAbsoluteRemoteFs(), remoteFS));
+            return createPath(computer.getAbsoluteRemoteFs() == null ? remoteFS : computer.getAbsoluteRemoteFs());
```
Comment (StefanSpieker) [inline core/src/main/java/hudson/model/Slave.java:395]:
```suggestion
            return createPath(Objects.toString(computer.getAbsoluteRemoteFs(), remoteFS));
```

File: core/src/main/java/hudson/tasks/BuildTrigger.java [hunk 3]
```diff
@@ -114,7 +113,7 @@ public BuildTrigger(String childProjects, boolean evenIfUnstable) {
 
     @DataBoundConstructor
     public BuildTrigger(String childProjects, String threshold) {
-        this(childProjects, Result.fromString(StringUtils.defaultString(threshold, Result.SUCCESS.toString())));
+        this(childProjects, Result.fromString(threshold == null ? Result.SUCCESS.toString() : threshold));
```
Comment (StefanSpieker) [inline core/src/main/java/hudson/tasks/BuildTrigger.java:116]:
```suggestion
        this(childProjects, Result.fromString(Objects.toString(threshold, Result.SUCCESS.toString())));
```

**Review by StefanSpieker — COMMENTED** [review]
I briefly reviewed this PR, I'm not sure if all replacements are equivalent. At least the `StringUtils.defaultString` should be replaced by the suggestion - not sure that all imports still fit

File: cli/src/main/java/hudson/cli/CLI.java (hunk 1 above)
Comment (BobDu) [reply cli/src/main/java/hudson/cli/CLI.java:248]: I modified it in two other places.

However, Considering the context here, it is located within the `if` block of `!= null`. So there's no need to wrap it with `defaultString`.

File: core/src/main/java/hudson/cli/CLIAction.java [hunk 4]
```diff
@@ -128,7 +127,17 @@ public HttpResponse doWs(StaplerRequest req) {
...
+            String o = Jenkins.get().getRootUrlFromRequest();
+            String removeSuffix1 = "/";
+            if (o.endsWith(removeSuffix1)) {
+                o = o.substring(0, o.length() - removeSuffix1.length());
+            }
+            String removeSuffix2 = req.getContextPath();
+            if (o.endsWith(removeSuffix2)) {
+                o = o.substring(0, o.length() - removeSuffix2.length());
+            }
+            final String expectedOrigin = o;
```
Comment (basil) [inline core/src/main/java/hudson/cli/CLIAction.java:140]: The extra `o` variable seems like an unnecessary layer of indirection for the reader. Why not just name it `expectedOrigin` to begin with and remove this line.

File: core/src/main/java/hudson/PluginManager.java [hunk 5]
```diff
@@ -1436,14 +1435,14 @@ public HttpResponse doPluginsSearch(@QueryParameter String query, @QueryParamete
...
+                    return (plugin.name != null && plugin.name.toLowerCase().contains(query.toLowerCase())) ||
+                        (plugin.title != null && plugin.title.toLowerCase().contains(query.toLowerCase())) ||
+                        (plugin.excerpt != null && plugin.excerpt.toLowerCase().contains(query.toLowerCase())) ||
                         plugin.hasCategory(query) ||
                         plugin.getCategoriesStream()
                             .map(UpdateCenter::getCategoryDisplayName)
-                            .anyMatch(category -> StringUtils.containsIgnoreCase(category, query)) ||
-                        plugin.hasWarnings() && query.equalsIgnoreCase("warning:");
+                            .anyMatch(category -> category != null && category.toLowerCase().contains(query.toLowerCase())) ||
+                        (plugin.hasWarnings() && query.equalsIgnoreCase("warning:"));
```
Comment (basil) [inline core/src/main/java/hudson/PluginManager.java:1445]: Seems like an unrelated change which could be removed to shrink the size of the diff and therefore its readability.

File: core/src/main/java/hudson/cli/CLIAction.java (hunk 4 above)
Comment (BobDu) [reply core/src/main/java/hudson/cli/CLIAction.java:140]: `expectedOrigin` is a local variable used in lambda, it must be a `final` or `effectively final`.

Of course we can set `o` as another name for readability, but I think it is not very necessary as a temporary variable.

File: core/src/main/java/hudson/PluginManager.java (hunk 5 above)
Comment (BobDu) [reply core/src/main/java/hudson/PluginManager.java]: already revert it.

File: core/src/main/java/hudson/model/AutoCompletionCandidates.java [hunk 6]
```diff
@@ -173,6 +172,6 @@ private String contextualNameOf(Item i) {
     }
 
     private static boolean startsWithImpl(String str, String prefix, boolean ignoreCase) {
-        return ignoreCase ? StringUtils.startsWithIgnoreCase(str, prefix) : str.startsWith(prefix);
+        return ignoreCase ? str.toLowerCase().startsWith(prefix.toLowerCase()) : str.startsWith(prefix);
```
Comment (basil) [inline core/src/main/java/hudson/model/AutoCompletionCandidates.java:175]: I couldn't easily prove to myself that `prefix` is always non-null here. Am I missing something? If not, then I think we should handle a null `prefix` to minimize risk of regression.

File: core/src/main/java/hudson/search/FixedSet.java [hunk 7]
```diff
@@ -62,7 +61,7 @@ public void suggest(String token, List<SearchItem> result) {
         boolean caseInsensitive = UserSearchProperty.isCaseInsensitive();
         for (SearchItem i : items) {
             String name = i.getSearchName();
-            if (name != null && (name.contains(token) || (caseInsensitive && StringUtils.containsIgnoreCase(name, token)))) {
+            if (name != null && (name.contains(token) || (caseInsensitive && name.toLowerCase().contains(token.toLowerCase())))) {
```
Comment (basil) [inline core/src/main/java/hudson/search/FixedSet.java:64]: Ditto

**Review by basil — COMMENTED** [review]
Looks great! Many thanks for this contribution. Just two minor comments and I am ready to approve this PR.

File: core/src/main/java/hudson/model/AutoCompletionCandidates.java (hunk 6 above)
Comment (BobDu) [reply core/src/main/java/hudson/model/AutoCompletionCandidates.java:175]: we already use `str.startsWith(token)` when `ignoreCase` equals `false`.
`startsWith` method args value must not null.

File: core/src/main/java/hudson/search/FixedSet.java (hunk 7 above)
Comment (BobDu) [reply core/src/main/java/hudson/search/FixedSet.java:64]: `String.contains` method args value must not null, too.

**Review by basil — APPROVED** [review]
Many thanks for the PR!

**Review by MarkEWaite — APPROVED** [review]
This PR is now ready for merge.  We will merge it after approximately 24 hours if there is no negative feedback.

/label ready-for-merge

File: core/src/main/java/jenkins/security/BasicHeaderProcessor.java [hunk 8]
```diff
@@ -65,7 +64,7 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
         HttpServletResponse rsp = (HttpServletResponse) response;
         String authorization = req.getHeader("Authorization");
 
-        if (StringUtils.startsWithIgnoreCase(authorization, "Basic ")) {
+        if (authorization != null && authorization.toLowerCase().startsWith("Basic ".toLowerCase())) {
```
Comment (jglick) [inline core/src/main/java/jenkins/security/BasicHeaderProcessor.java:67]: While it is unlikely to matter in practice, this is _not_ a sound replacement. If a request was sent with

```
AUTHORIZATION: BASIC czNjcjN0
```

then a server running in Turkish locale would fail this clause because

```java
"basıc cznjcjn0".startsWith("basic ")
```

is false.

In general, you always want to pass the `Locale` argument (almost always `Locale.ROOT`) to `toLowerCase` / `toUpperCase`.

Comment (basil) [reply core/src/main/java/jenkins/security/BasicHeaderProcessor.java:67]: @BobDu Is it worth filing a follow-up PR to address the above feedback? Or do we think the code is fine as-is?

Comment (BobDu) [reply core/src/main/java/jenkins/security/BasicHeaderProcessor.java:67]: I agree always add `Locale.ROOT` arg to `toLowerCase` method is a best practices.
Already make a new PR #9162
